[HCP Observability] OTELSink

Achooo opened the pull request:
### Description

<!-- Please describe why you're making this change, in plain English. -->
##### Overall Feature background
This PR is part of the HCP Observability feature to send Server metrics to HCP's new Telemetry Gateway. The idea is to create a new OpenTelemetry go-metrics sink that will be initialized automatically when a Consul server is registered with the HCP management plane.
- We will be merging these approved changes into **feature/hcp-telemetry**
- That branch will be merged at the end into `main` and back ported
This allows us to iteratively improve on this feature over time before. Split up for easier PR review. 🚢 

##### This PR: MetricsClient
This PR adds a new `OTELSink` to the new Telemetry package.
It must satisfy the [go metrics MetricsSink](https://github.com/hashicorp/go-metrics/blob/8cabd9eab1bee469056110d091281261e861fc19/sink.go#L10) interface to be used by the fanout sink (follow up PR).
The key responsibility of the OTELExpoter is to use the MetricsClient to export metrics. To do this:
- Receives an emitted Consul metric 
- Registers a metric with an OTEL instrument
- Periodically aggregates and exports data using a PeriodicReader. This reader uses the Exporter implemented in #17128 

The OTEL library pieces are pegged at this commit hash `7dea2225a218872e86d2f580e82c089b321617b0` version due to build failures on certain architectures. See this build [failure](https://github.com/hashicorp/consul/actions/runs/4981525713/jobs/8915904823?pr=17162).

See this GitHub [issue](https://github.com/open-telemetry/opentelemetry-go/issues/4076):

Although the fixes were implemented, this has yet to be released.
As soon as the fix is released, this version will be updated.

##### Next steps
- The sink will be initialized when a consul server is registered with CCM
- The initialized sink will be registered in lib/telemetry.go's fanout sinks

##### Reference image

![minirfc_optionA_otelsink drawio-8](https://user-images.githubusercontent.com/23405655/234686584-9fc906fd-df48-4db6-b5dc-bacb46ece482.png)

### Testing & Reproduction steps

<!--

* In the case of bugs, describe how to replicate
* If any manual tests were done, document the steps and the conditions to replicate
* Call out any important/ relevant unit tests, e2e tests or integration tests you have added or are adding

-->

### Links
<!--

Include any links here that might be helpful for people reviewing your PR (Tickets, GH issues, API docs, external benchmarks, tools docs, etc). If there are none, feel free to delete this section.

Please be mindful not to leak any customer or confidential information. HashiCorp employees may want to use our internal URL shortener to obfuscate links.

-->

### PR Checklist

* [x] updated test coverage

* [ ] external facing docs updated
* [ ] appropriate backport labels added (not yet, in feature branch)
* [ ] not a security concern

Achooo commented on agent/hcp/telemetry/gauge_store.go:
I could pretty much remove this whole thing with the sync.Map implementation, see this commit : https://github.com/hashicorp/consul/pull/17159/commits/e9f862d6e61db93b50be546819cc11de2cfc701f


I had to do this to synchronize access to the global store. I think we could shift back to sync.Map if we prefer to keep the code brief and arguable safer (no bugs)

The reason I shifted away from sync.Map is because of all the type assertions

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,198 @@
+	meterProvider *otelsdk.MeterProvider
+	meter         *otelmetric.Meter
+
+	gaugeInstruments     map[string]*instrument.Float64ObservableGauge
```
With sync.Map the types aren't very obvious, this felt easier to read.
However, this means we need to use a mutex.

All my operations require a write, so I used `sync.Mutex` instead of RWMutex but open to thoughts.

Achooo commented on agent/hcp/telemetry/gauge_store.go:
```diff
@@ -0,0 +1,75 @@
+
+type gaugeStore struct {
+	store map[string]*gaugeValue
+	mutex sync.Mutex
```
All my operations require a write, so I used `sync.Mutex` instead of RWMutex (since the store also deletes).

loshz replied:
Then I would say your "read" is also a write. Maybe extract the reading part and add a "delete" method?

Achooo replied:
I never need to read only, as we need to always delete the value after reading. I'll also update the code to return early if the value doesn't exist. 

I feel like having it in a single method with a single write mutex makes it more atomic and simpler? By separating these, the callback will have to call read, then delete separately. Especially, as the gauge store won't be used anywhere else (I assume).

Is there a reason to extract both and prefer RWMutex?

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,198 @@
+
+// gaugeCallback returns a callback which gets called when metrics are collected for export.
+// the callback obtains the gauge value from the global gauges.
+func gaugeCallback(key string) instrument.Float64Callback {
```
ASH TODO: Move this onto the Gauge store and init within the OTELSink

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+type OTELSinkOpts struct {
```
Keeping this struct as we have one last follow up PR that adds default filters and labels to the sink (so we'll have 4 parameters)

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+
+type OTELSink struct {
+	spaceReplacer *strings.Replacer
+	logger        hclog.Logger
```
I'll modify this to use `hclog.FromContext`

chapmanc commented on agent/hcp/telemetry/gauge_store.go:
```diff
@@ -0,0 +1,61 @@
+package telemetry
```
Same as before please use a doc.go with a short explanation when you have a chance.

Achooo replied:
Added with comment about the sink in https://github.com/hashicorp/consul/pull/17128!

chapmanc commented on agent/hcp/telemetry/gauge_store.go:
```diff
@@ -0,0 +1,61 @@
+// the callback obtains the gauge value from the global gauges.
+func (g *gaugeStore) gaugeCallback(key string) instrument.Float64Callback {
+	// Closures keep a reference to the key string, that get garbage collected when code completes.
+	return func(_ context.Context, obs instrument.Float64Observer) error {
```
Considering we are passing in context that we should probably check if the context has been cancelled and return an error.


```suggestion
	return func(ctx context.Context, obs instrument.Float64Observer) error {
	
	select {
	case <-ctx.Done:
	   return error of some sort
	 default:
	}
```

chapmanc commented on agent/hcp/telemetry/gauge_store_test.go:
```diff
@@ -0,0 +1,44 @@
+)
+
+func TestGaugeStore(t *testing.T) {
+	globalGauges := &gaugeStore{
```
Maybe we could add a couple concurrency tests to validate any assumptions and to document your intent.

Achooo replied:
Will resolve this and tag you in the new concurrent tests!

chapmanc commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+
+type OTELSinkOpts struct {
+	Reader otelsdk.Reader
+	Logger hclog.Logger
```
TODO: use ctx

chapmanc commented on agent/hcp/telemetry/gauge_store.go:
```diff
@@ -0,0 +1,61 @@
+	"go.opentelemetry.io/otel/metric/instrument"
+)
+
+type gaugeStore struct {
```
Might be good to add some comments here as to why we need a gaugeStore and we don't need a counterStore/histogramStore

Achooo replied:
Added one, let me know if it can be improved!

chapmanc commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	}
+
+	if opts.Reader == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Reader")
```
If opts.Reader == nil can we default it to the OtelReader above?

Achooo replied:
We'd need an initialized MetricsClient to configure the exporter within the reader. We can default to a ManualReader (i.e. reader one used in tests).

chapmanc commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+
+	// If instrument does not exist, create it and register callback to emit last value in global Gauge store.
+	if _, ok := o.gaugeInstruments[k]; !ok {
+		inst, err := (*o.meter).Float64ObservableGauge(k, instrument.WithFloat64Callback(o.gaugeStore.gaugeCallback(k)))
```
Can we document that this call back is really just registering the gauge for the first time? After that it should be in the store.

chapmanc commented on agent/hcp/telemetry/otel_sink_test.go:
```diff
@@ -0,0 +1,250 @@
+
+		require.Equal(t, dp.Attributes.Len(), currActual.Attributes.Len(), "different attributes of datapoint length")
+
+		iter := dp.Attributes.Iter()
```
Why not use require.Equal?

Achooo replied:
yep! fixed up

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/agent/hcp/client"
+
+	gometrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"go.opentelemetry.io/otel/attribute"
+	otelmetric "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/instrument"
+	otelsdk "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
```
Nit:
```suggestion
	gometrics "github.com/armon/go-metrics"
	"github.com/hashicorp/go-hclog"
	"go.opentelemetry.io/otel/attribute"
	otelmetric "go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/metric/instrument"
	otelsdk "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	
	"github.com/hashicorp/consul/agent/hcp/client"
```

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	Logger hclog.Logger
+}
+
+type OTELSink struct {
```
This whole struct needs documenting.

Achooo replied:
I've added some docs, and tried to be quite detailed here! let me know if there are any wording changes we could do to make it better.

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+
+func NewOTELSink(opts *OTELSinkOpts) (*OTELSink, error) {
+	if opts.Logger == nil {
+		return nil, fmt.Errorf("failed to init OTEL sink: provide valid OTELSinkOpts Logger")
```
Personally, I wouldn't hardcode type names in error messages:
```suggestion
		return nil, fmt.Errorf("ferror: logger cannot be nil")
```

Achooo replied:
Yep thanks! Fixed it, will resolve but feel free to suggest changes on the messages again.

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+		logger:               opts.Logger.Named("otel_sink"),
+		meterProvider:        meterProvider,
+		meter:                &meter,
+		mutex:                sync.Mutex{},
```
You don't need to initialise a mutex in Go.
```suggestion
```

Achooo replied:
Thanks - good point ! fixed this

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+
+	gs := &gaugeStore{
+		store: make(map[string]*gaugeValue, 0),
+		mutex: sync.Mutex{},
```
You don't need to initialise a mutex in Go.
```suggestion
```

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	}
+
+	attrs := toAttributes(labels)
+	(*inst).Record(context.TODO(), float64(val), attrs...)
```
Should we be using a cancelable context here?

Achooo replied:
Yeah good question, so we have a timeout for the actual net/http request which happens at the export stage. 
In this block of code, the measurement of a metric gets recorded and added to the aggregation. The export happens every 10 seconds, and it's not related to this code block.

From my digging in the SDK, the context is passed here as an early return in case the ctx is cancelled. 
So if the ctx is cancelled, metrics won't get aggregated nor recorded, see source code: 

https://github.com/open-telemetry/opentelemetry-go/blob/24e8d2e5749d22fcd2266eb3e4d35bbe9d42259f/sdk/metric/instrument.go#L190C2-L202

In their usage [example](https://github.com/open-telemetry/opentelemetry-go/blob/24e8d2e5749d22fcd2266eb3e4d35bbe9d42259f/example/prometheus/main.go#L82.) in their official docs, they use a `context.Background` and await interrupt signals / ctx cancellation.

The right thing to do here would be to use a parent context, to prevent metric aggregation in the case where ctx is cancelled. However, we don't have that in the code right now since the go metrics sink methods don't pass in a context 😓 I feel like they should... A timeout could ensure this aggregation doesn't take too long - which it shouldn't. We could also end up missing metrics. In think for that, should we be more data driven and look at performance first to see if it's needed?

Achooo replied:
I was definitely unsure what to do here, so let me know if you have any preferences

loshz commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	}
+
+	attrs := toAttributes(labels)
+	(*inst).Add(context.TODO(), float64(val), attrs...)
```
Should we be using a cancelable context here?

Achooo replied:
same as https://github.com/hashicorp/consul/pull/17159#discussion_r1191361841

Achooo replied:
Resolving this and we can discuss in that thread!

Achooo commented on agent/hcp/telemetry/otel_sink.go:
```diff
@@ -0,0 +1,189 @@
+	logger        hclog.Logger
+
+	meterProvider *otelsdk.MeterProvider
+	meter         *otelmetric.Meter
```
I've added some docs, and tried to be quite detailed here! let me know if there are any wording changes we could do to make it better.

Achooo commented on agent/hcp/telemetry/gauge_store.go:
```diff
@@ -0,0 +1,61 @@
+
+	gauge, ok := g.store[key]
+
+	delete(g.store, key)
```
yep thanks 😓

Achooo commented on agent/hcp/telemetry/gauge_store_test.go:
```diff
@@ -0,0 +1,81 @@
+
+// TestGaugeStore_Race induces a race condition. When run with go test -race,
+// this test should pass if  implementation is concurrency safe.
+func TestGaugeStore_Race(t *testing.T) {
```
@chapmanc new test using wait groups, which performs concurrent operations. The test suite also runs tests using `go -race` so if these pass, assumptions are validated.

Achooo commented on agent/hcp/telemetry/otel_sink_test.go:
```diff
@@ -0,0 +1,282 @@
+		},
+	}
+
+	wg := &sync.WaitGroup{}
```
@chapmanc converted this test to use wait group and perform instrumentation concurrently to test concurrency assumptions.

Achooo replied:
I'll update this test to do something similar to what's done in the gauge store

chapmanc commented on agent/hcp/telemetry/gauge_store_test.go:
```diff
@@ -0,0 +1,81 @@
+
+	gaugeStore := NewGaugeStore()
+	wg := &sync.WaitGroup{}
+	for k, v := range map[string]float64{"consul.raft.apply": 23.23, "consul.raft.test": 14.3} {
```
Could we increase the test size from two samples ? Would a function that would generate 100 k,v would give us a better representation ? Otherwise two competing gorountines isn't likely to reveal much about the concurrency issues this might face.

Achooo replied:
ah yeah I was wondering about that and fuzz testing but decided to keep it simple. I'll update to 100 generated kvs

chapmanc replied:
https://go.dev/blog/race-detector

> Because of its design, the race detector can detect race conditions only when they are actually triggered by running code, which means it’s important to run race-enabled binaries under realistic workloads. 

I might be wrong but I thought you still needed realistic workloads.

Achooo replied:
hmm that makes sense actually, yeah but still more realistic than 2 🤔 I'll update the PR, thanks for the context. I was misunderstanding the way it works I think!

loshz replied:
>which should theoretically be able to tell from just 2 operations

As @chapmanc said, unfortunately 2 ops is significantly less likely to perform a race due to the optimizations the OS makes at runtime. Using 100+ random generated results would be more reliable.

Achooo replied:
thank you both, TIL. I'm curious and will look more into the race detector for fun too. :D I updated the PR to have 100 samples for the gauge kv

Achooo commented on agent/hcp/telemetry/gauge_store_test.go:
```diff
@@ -0,0 +1,88 @@
+
+	gaugeStore := NewGaugeStore()
+	wg := &sync.WaitGroup{}
+	for i := 0; i < 100; i++ {
```
@chapmanc wdyt of this new version with 100 generated kv pairs? uuid is prob unique enough but I also combine the math.rand value in the key to make it even better.

Achooo replied:
the one downside with this approach is that we have to generate these and guarantee uniqueness

chapmanc replied:
I think I am good with this. This mimics actual use.

Achooo replied:
I'm a dummy and updated the PR to use a kv store that generates unique keys and names in a better way without need for uuid...

Achooo commented:
@chapmanc @loshz thank you both for the thorough feedback! I think I've addressed all the comments, and I hope I didn't miss anything. Would love your re-review 🙏🏽 

A remaining question is about the context when recording a measurement, see this [thread](https://github.com/hashicorp/consul/pull/17159#discussion_r1188450219).

loshz commented on agent/hcp/telemetry/gauge_store_test.go:
```diff
@@ -0,0 +1,87 @@
+func TestGaugeStore_Race(t *testing.T) {
+	t.Parallel()
+
+	gaugeStore := NewGaugeStore()
+	wg := &sync.WaitGroup{}
+	samples := 100
+	for i := 0; i < samples; i++ {
+		k := fmt.Sprintf("consul.test.%d", i)
+		v := rand.Float64()
+		wg.Add(1)
+		go storeAndRetrieve(t, k, v, gaugeStore, wg)
+	}
+
+	wg.Wait()
+}
+
+func storeAndRetrieve(t *testing.T, k string, v float64, gaugeStore *gaugeStore, wg *sync.WaitGroup) {
+	gaugeStore.Set(k, v, nil)
+	gv, ok := gaugeStore.LoadAndDelete(k)
+	require.True(t, ok)
+	require.Equal(t, v, gv.Value)
+	wg.Done()
+}
```
This approach is definitely more robust when testing for race conditions!

Small suggestion, I would change the goroutines to not pass the WaitGroup, and also not use a random value in order to save compute:
```suggestion
	gaugeStore := NewGaugeStore()
	wg := &sync.WaitGroup{}
	for i := 0; i < 100; i++ {
		wg.Add(1)
		key := fmt.Sprintf("consul.test.%d", i)
		value := 12.34
		go func() {
			defer wg.Done()
			
			gaugeStore.Set(key, value, nil)
			gv, ok := gaugeStore.LoadAndDelete(k)
			require.True(t, ok)
			require.Equal(t, value, gv.Value)
		}()
	}

	wg.Wait()
}
```

loshz replied:
I'm also unsure how the `require` package would work in this situation. I _think_ it fails the test immediately, so I'm not 100% what will happen to the `wg.Wait()` :thinking:

loshz replied:
Yep, it calls [`t.FailNow()`](https://github.com/stretchr/testify/blob/f36bfe3c337aa95c86f04c721acdbafb5ffb1611/require/require.go#L1825) under the hood which is not safe inside a goroutine.

I would create a buffered channel before the loop, send errors in the goroutines, and assert no errors after the call to `wg.Wait()`. Something like this should do it:
```go
func TestGaugeStore_Race(t *testing.T) {
	t.Parallel()

	gaugeStore := NewGaugeStore()
	wg := &sync.WaitGroup{}
	samples := 100
	errCh := make(chan error, samples)

	for i := 0; i < samples; i++ {
		wg.Add(1)
		key := fmt.Sprintf("consul.test.%d", i)
		value := 12.34
		go func() {
			defer wg.Done()
			gaugeStore.Set(key, value, nil)
			gv, _ := gaugeStore.LoadAndDelete(k)
			if gv.Value != value {
				errCh <- fmt.Errorf("expected value: '%s', but got: '%s' for key: '%s'", value, gv.Value, key)
			}
		}()
	}
	wg.Wait()
}

Achooo replied:
thanks for catching this, I've fixed it and for the OTELSink test too.

loshz commented on agent/hcp/telemetry/gauge_store.go:
```diff
@@ -0,0 +1,77 @@
+// LoadAndDelete will read a Gauge value and delete it.
+// Once registered for a metric name, a Gauge callback will continue to execute every collection cycel.
+// We must delete the value once we have read it, to avoid repeat values being sent.
+func (g *gaugeStore) LoadAndDelete(key string) (*gaugeValue, bool) {
```
Would the caller ever care if the value is not in the `gaugeStore`? Trying to determine if we need to return the ok bool?

Achooo replied:
I check if there is a value to emit here:

https://github.com/hashicorp/consul/pull/17159/files#diff-2f8b324f4e5da8b8ad5221a272ca430aa377e759443fa347f33b6524906661c4R71-R73

Achooo replied:
thanks for catching this, I've fixed it and for the otel test too.

loshz commented on agent/hcp/telemetry/otel_sink_test.go:
```diff
@@ -0,0 +1,386 @@
+	for k, v := range expectedMetrics {
+		wg.Add(1)
+		go func(k string, v metricdata.Metrics) {
+			performSinkOperation(t, sink, k, v, errCh)
+			wg.Done()
```
Nit:
```suggestion
			defer wg.Done()
			performSinkOperation(t, sink, k, v, errCh)
```

loshz commented on agent/hcp/telemetry/otel_sink_test.go:
```diff
@@ -0,0 +1,386 @@
+}
+
+// performSinkOperation emits a measurement using the OTELSink and calls wg.Done() when completed.
+func performSinkOperation(t *testing.T, sink *OTELSink, k string, v metricdata.Metrics, errCh chan error) {
```
I don't think we need `t` anymore? 
```suggestion
func performSinkOperation(sink *OTELSink, k string, v metricdata.Metrics, errCh chan error) {
```

loshz reviewed (approved):
Really great work, @Achooo! Thanks for fixing up the concurrency tests. I've left a couple of nits, but I'll approve as to not require another review.
